hadoop-crypto/src/test/java/com/palantir/hadoop/ConfigurableEncryptedFileSystemTest.java

```diff
@@ -0,0 +1,171 @@
+        IOUtils.readFully(dis, readData);
+        assertThat(readData, is(dataBytes));
+
+        // Raw data is not the same
+        dis = rawFs.open(path);
+        IOUtils.readFully(dis, readData);
+        assertThat(readData, is(not(dataBytes)));
+
+        // KeyMaterial file exists
+        assertTrue(rawFs.exists(new Path(path + FileKeyStorageStrategy.EXTENSION)));
+    }
+
+    @Test
+    public void testOnlyPublicKey() throws IOException {
+        String data = "data";
+        byte[] dataBytes = data.getBytes(StandardCharsets.UTF_8);
+        byte[] readData = new byte[data.length()];
+
+        conf.unset(ConfigurableEncryptedFileSystem.PRIVATE_KEY_CONF);
+        FileSystem efsPublic = FileSystem.newInstance(EFS_URI, conf);
+
+        // Write encrypted data
+        OutputStream os = efsPublic.create(path);
+        IOUtils.write(data, os);
+        os.close();
+
+        // Raw data is not the same
+        InputStream dis = rawFs.open(path);
+        IOUtils.readFully(dis, readData);
+        assertThat(readData, is(not(dataBytes)));
+
+        // KeyMaterial file exists
+        assertTrue(rawFs.exists(new Path(path + FileKeyStorageStrategy.EXTENSION)));
+
+        // Unable to open files without Private Key
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("Private key is absent but required to get key material");
+        efsPublic.open(path);
+    }
+
+    @Test
+    public void testNoPublicKey() throws IOException {
+        expectedException.expect(NullPointerException.class);
+        expectedException.expectMessage(String.format("Public Key must be configured for key %s",
+                ConfigurableEncryptedFileSystem.PUBLIC_KEY_CONF));
+        FileSystem.newInstance(EFS_URI, getBaseConf());
+    }
+
+    @Test
+    public void testBackingFsInvalid() throws IOException {
+        conf = getBaseConf();
+        conf.set("fs.nope.impl", ConfigurableEncryptedFileSystem.class.getCanonicalName());
+
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("URI scheme must begin with 'e' but received: nope");
+        FileSystem.newInstance(URI.create("nope:///"), conf);
+    }
+
+    @Test
+    public void testFileNameCollidesWithKeyMaterial() throws IOException {
+        String data = "data";
+
+        // Write encrypted data
+        OutputStream os = efs.create(path);
+        IOUtils.write(data, os);
+        os.close();
+
+        // Write encrypted data
+        os = efs.create(new Path(path + FileKeyStorageStrategy.EXTENSION));
+        IOUtils.write(data, os);
+        os.close();
+
+        // NOTE(jellis): This is the most likely exception, however if the first byte of the .keymaterial file is the
+        // same as the KeyMaterials version then any number of RuntimeExceptions may be thrown.
+        // expectedException.expect(IllegalArgumentException.class);
```

Comment:
the goal of the test is to say if someone writes a file ending in `.keymaterial` then the EFS fails to decrypt the file whos keymaterial has been overwritten. you'd expect it to fail due to incorrect format version most of the time but it's possible for the first encrypted byte of the new file to be 1 or 2 which would cause the EFS to fail for any number of reasons. may not even be worth this comment though.